addons/website_sale/models/sale_order.py

```diff
@@ -1,59 +1,55 @@
 # -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
 import random
-import openerp
 
-from openerp import SUPERUSER_ID, tools
-import openerp.addons.decimal_precision as dp
-from openerp.osv import osv, orm, fields
-from openerp.addons.web.http import request
-from openerp.tools.translate import _
-from openerp.exceptions import UserError
+from odoo import api, models, fields, tools, _
+from odoo.http import request
+from odoo.exceptions import UserError
+import odoo.addons.decimal_precision as dp
 
 
-class sale_order(osv.Model):
+class SaleOrder(models.Model):
     _inherit = "sale.order"
 
-    def _cart_info(self, cr, uid, ids, field_name, arg, context=None):
-        res = dict()
-        for order in self.browse(cr, uid, ids, context=context):
-            res[order.id] = {
-                'cart_quantity': int(sum(l.product_uom_qty for l in (order.website_order_line or []))),
-                'only_services': all(l.product_id and l.product_id.type in ('service', 'digital') for l in order.website_order_line)
-            }
-        return res
-
-    _columns = {
-        'website_order_line': fields.one2many(
-            'sale.order.line', 'order_id',
-            string='Order Lines displayed on Website', readonly=True,
-            help='Order Lines to be displayed on the website. They should not be used for computation purpose.',
-        ),
-        'cart_quantity': fields.function(_cart_info, type='integer', string='Cart Quantity', multi='_cart_info'),
-        'payment_acquirer_id': fields.many2one('payment.acquirer', 'Payment Acquirer', on_delete='set null', copy=False),
-        'payment_tx_id': fields.many2one('payment.transaction', 'Transaction', on_delete='set null', copy=False),
-        'only_services': fields.function(_cart_info, type='boolean', string='Only Services', multi='_cart_info'),
-    }
-
-    def _get_errors(self, cr, uid, order, context=None):
+    website_order_line = fields.One2many(
+        'sale.order.line', 'order_id',
+        string='Order Lines displayed on Website', readonly=True,
+        help='Order Lines to be displayed on the website. They should not be used for computation purpose.',
+    )
+    cart_quantity = fields.Integer(compute='_compute_cart_info', string='Cart Quantity')
+    payment_acquirer_id = fields.Many2one('payment.acquirer', string='Payment Acquirer', copy=False)
+    payment_tx_id = fields.Many2one('payment.transaction', string='Transaction', copy=False)
+    only_services = fields.Boolean(compute='_compute_cart_info', string='Only Services')
+
+    @api.multi
+    def _compute_cart_info(self):
+        for order in self:
+            order.cart_quantity = int(sum(order.mapped('website_order_line.product_uom_qty')))
+            order.only_services = not bool(order.website_order_line.filtered(lambda line: line.product_id.type not in ('service', 'digital')))
```

Comment:
I don't see why you changed the previous code that was quite clear `all` instead of a double-negations `not bool(... not in)`